# Primeros pasos con Python

Después de hacer algunos programas en Turtle donde se han visto algunos principios de programación, muchos de los relevantes, ahora haremos una introducción a la programación en general. 

## 1. Ejecutando  Python

Uno de los desafíos al comenzar a trabajar con Python es instalar el software relacionado en tu computadora.

Si estás familiarizado con su sistema operativo, y especialmente si te sentís cómodo con la interfaz de línea de comandos, no vas a tener problemas para instalar Python. Pero para los principiantes, puede ser complicado  aprender sobre administración del sistema y programación al  mismo tiempo.

Para evitar ese problema, recomendamos comenzar con Jupyter Notebooks. *Jupyter Notebook* (anteriormente IPython Notebooks) es un entorno informático que permite trabajar alternadamente con programas Python y texto enriquecido. Este que están viendo es un *cuaderno Jupyter*.  

Existen varias implementaciones de cuadernos Jupyter para comezar a trabajar con ellas:

- *Google Colaboratory (Colab).* Es una versión de Jupyter Notebooks on line provista por Google. Es una fabulosa herramienta didáctica y nos permitirá hacer gran parte del curso aunque en  las últimas semanas posiblemente necesitemos otra herramienta. Las tareas y ejercicios del curso se harán con esta herramienta.
- *Anaconda.* Anaconda es mucho más que Jupyter Notebooks y los que se sientan cómodos con esta aplicación podrán utilizarla en el curso.
- *JupyterLab Desktop.*  Es un subproducto de Anaconda, pero mucho más fácil de instalar y manejar.
- *Visual Studio Code + Python.* Quizás la forma más conveniente y cómoda de manejar cuadernos de Jupyter y scripts de Python puro en un solo entorno. Es altamente recomendable y  encontrarás instruciones de como instalar este entorno en la Web o preguntando a alguna IA. Conceptualmente debes seguir los siguientes pasos:
    - Instalar Python 3.
    - Instalar VSCode (Visual Studio  Code). Es de código abierto y libre uso.
    - Instalar extensiones de VSCode para poder incorporar y ejecutar cuadernos Jupyter.

Un  cuaderno Jupyter se divide en celdas de texto con formato y celdas de código Python. Cuando creamos una celda elegimos que tipo de celda queremos ("Markdown" para texto, "Code" para Python). Obviamente lo que estamos leyendo ahora es una celda de texto. La celda siguiente es una celda de código y la podemos ejecutar cliqueando el triangulito o posicionarse en ella  y hacer `Mayus+Enter`:




In [1]:
1 + 1

2

En muchos proyectos que involucren desarrollo de software en Python se utiliza *el modo de secuencia de comandos* o *modo script*. En este modo cada pieza de código se guarda en un archivo y luego se ejecuta este archivo, llamado *script*, en el intérprete. Por convención, los scripts de Python tienen nombres que terminan en `.py`. Si sabés cómo crear y ejecutar un script en tu computadora podrás correr todos los scripts de este curso.

Hay herramientas muy útiles para escribir código llamadas IDE (Integrated Development Environment) o entornos de desarrollo integrado​s. Una IDE es una aplicación informática que proporciona servicios integrales para facilitarle al desarrollador o programador el desarrollo de software.

La primera que ya  mencionamos es Visual Studio Code. Visual Studio Code es un editor de código fuente desarrollado por Microsoft para Windows, Linux, macOS y Web. Incluye soporte para la depuración, control integrado de Git, resaltado de sintaxis, finalización inteligente de código, fragmentos y refactorización de código. También es personalizable, por lo que los usuarios pueden cambiar el tema del editor, los atajos de teclado y las preferencias. Es gratuito y de código abierto. Por medio de plugins es posible codificar en casi cualquier lenguaje de programación, en particular en Python.

La otra herramienta a mencionar es PyCharm. PyCharm es un IDE multiplataforma dirigido principalmente a Python y al desarrollo web. Una versión de código abierto está disponible como PyCharm Community Edition. Hay una versión paga llamada PyCharm Professional Edition. Una versión especial llamada PyCharm Edu está basada en PyCharm, dirigida específicamente al aprendizaje de la programación con Python. Como alumno universitario podés aplicar a la versión académica. PyCharm tiene una distribución para Anaconda que se puede instalar junto con Anaconda y están estrechamente integradas.

En la primer celda que ejecutamos usamos a Python como calculadora y ese es el uso  mas sencillo que le podemos dar. Profundizaremos más abajo en este aspecto.

Ahora vemos algunos ejemplos muy sencillos de scripts:

In [2]:
print('¡Hola, mundo!')

¡Hola, mundo!


Este es un ejemplo de una _sentencia print_, aunque en realidad no imprime nada en papel. Muestra un resultado en la pantalla. En este caso, como ya habrás comprobado, el resultado son las palabras
```
¡Hola Mundo!
```
Las comillas en el programa,  que pueden ser simples o dobles, marcan el principio y el final del texto que se mostrará y no aparecen en el resultado.

Los paréntesis indican que `print` es una _función_. Veremos funciones con profundidad más adelante.


In [4]:
print('El "Tata" Martino')
print("El 'Tata' Martino")
print("El \"Tata\" Martino")


El "Tata" Martino
El 'Tata' Martino
El "Tata" Martino


Alternando  comillas simple y comillas dobles o utilizando `\'` y `\"` podemos imprimir comillas.

## 2. Operadores aritméticos

Después de `Hola, mundo`, el siguiente paso es la aritmética elemental. Python proporciona _operadores_, que son símbolos especiales que representan cálculos
como suma y multiplicación.

Los operadores `+`, `-` y `*` realizan sumas, resta y multiplicación, respectivamente. Por ejemplo:


In [5]:
40 + 2

42

In [6]:
43 - 1

42

In [7]:
6 * 3

18

El operador `/` realiza la división:



In [8]:
84 / 2

42.0

Quizás te preguntes al ejecutar la celda anterior por qué el resultado es `42.0` en lugar de `42`. La respuesta es que Python tiene dos operadores de división. El primero,  es el que ya vimos, es `/` que siempre obtiene  un número decimal o coma flotante. El segundo es `//` y es la _división entera_. Por ejemplo,


In [9]:
84 // 2

42

devuelve `42` y

In [10]:
20 // 3

6

In [11]:
20/6

3.3333333333333335

Devuelve `6`, pues "3 entra 6 veces en 20".

Finalmente, el operador `**` realiza una exponenciación; esto es,
eleva un número a una potencia:

In [12]:
6 ** 2

36

Sabemos que $a^n$ es $a$ multiplicado por si mismo $n$-veces. También conocemos, que $(a^n)^m = a^{nm}$.

Entonces, como sabemos que si $a>0$, $(\sqrt[n]{a})^n = a$, es "claro" que $\sqrt[n]{a} = a^{1/n}$, pues $ (a^{1/n})^n = a^{\frac1n n} =a$.

Luego
$$
\sqrt{10} = 10^{1/2} = 10^{0.5}.
$$
Por ejemplo,

In [13]:
print(9**0.5) # raíz cuadrada de 9 (= 3)
print(27**(1/3)) # raíz cúbica de 27 (= 3)

3.0
3.0


En  las celdas de código anteriores usamos el símbolo `#` para comenzar un *comentario*: el texto que sigue después de `#` no será interpretado  o ejecutado y sirve de ayuda al programador.

En algunos lenguajes de programación se utiliza `^` para la exponenciación, pero en Python `^` es un operador bit a bit llamado XOR. Si no estás familiarizado con los operadores bit a bit, el resultado te sorprenderá:




In [14]:
6 ^ 2

4

No cubriremos los operadores bit a bit en este curso, pero podes leer acerca de
ellos en [http://wiki.python.org/moin/BitwiseOperators](http://wiki.python.org/moin/BitwiseOperators) (en inglés).

## 3. Valores y tipos

Un _valor_ es una de las entidades básicas con las que trabaja un programa, como una letra o un número. Algunos valores que hemos visto hasta ahora son `2`, `42.0` y `'¡Hola, mundo!'`.

Estos valores pertenecen a diferentes _tipos_: `2` es un _entero_, `42.0` es un _número de coma flotante_ y `'¡Hola, mundo!'`es una _cadena_, así llamada porque los caracteres que contiene se ubican en forma consecutiva.

Si no estás seguro de qué tipo de valor es un valor determinado, Python mismo puede decírtelo ejecutando la función `type()`.


In [15]:
print(type(2))
print(type(42.0))
print(type('¡Hola, mundo!'))
print(type(True))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


In [16]:
n = 2
print( n < 5)

True


Vemos en los resultados de ejecutar la celda anterior que en todos aparece la palabra `class`. La palabra `class`,  cuya traducción al español es *clase*, se usa en el sentido de una categoría: un tipo es una categoría de valores.

No es sorprendente que los enteros pertenezcan al tipo `int`, las cadenas pertenezcan a `str` (por _string_ en inglés) y los números en coma flotante pertenezcan a `float`.

El tipo `bool` solo tiene dos valores `True` y `False`,  que sirven para denotar la verdad o falsedad de expresiones lógicas.

¿Qué pasa con valores como `'2'` y `'42.0'`?
Parecen números, pero están entre comillas como cadenas y, efectivamente son cadenas. Lo  comprobaremos ejecutando lo siguiente:


In [ ]:
print(type('2'))
print(type('42.0'))

Cuando escribís un entero grande, podés tener la tentación de usar puntos entre grupos de dígitos, como en `1.000.000`. Sin embargo,  esto no es una forma permitida de escribir un entero, por ejemplo, si descomentamos la siguiente celda de código la ejecución nos devuelve un error:

In [ ]:
## print(1.000.000)

La forma correcta de escribir "un millón" es `1000000`. Python permite utilizar el guion bajo  en un entero para simplificar la lectura:  `1_000_000` es otra forma de escribir "un millón".

Podemos tratar también de escribir un número decimal o coma flotante usando la notación latina. Por ejemplo, "cuatro coma cinco", podríamos intentar escribirlo como `4,5`,  y hagamos la prueba:  

In [17]:
print(4,5)

4 5


¡Eso no es lo que esperábamos en absoluto! Python interpreta la expresión `4,5` como dos enteros separados.

Este es el primer ejemplo que hemos visto de un *error semántico*: el código se ejecuta sin producir un mensaje de error, pero no es lo que queremos obtener.

La forma correcta de escribir un número decimal es reemplazando  la coma por  el punto, por ejemplo "cuatro coma cinco" se escribe `4.5`

In [18]:
print(4.5 * 2)

9.0


## 4. Python como calculadora

Como vimos más arriba Python puede ser usado como una calculadora. Ahora veremos más funciones de "calculadora". Para ello utilizaremos el módulo `math`. El módulo `math` permite utilizar funciones matemáticas que no son accesibles directamente en Python. La instrucción para importar el módulo es

In [19]:
import math

y debemos ejecutarla si en las siguientes celdas de código queremos usar el módulo.

La instrucción para obtener el logaritmo en base $e$ es `math.log(t)`, donde `t` es un número real positivo.

Recordemos que si $a, t >0$ ,
$$
\operatorname{log}_a(t) = r \quad ⇔ \quad a^r = t
$$

In [20]:
math.log(10)

2.302585092994046

Si queremos obtener el logaritmo en base 10 debemos usar la función `math.log10()`:

In [21]:
print(math.log10(10))
print(math.log10(10**234))

1.0
234.0


Las funciones trigonométricas también son accesibles  desde el  módulo `math`.

In [22]:
math.sin(0)

0.0

es el seno  de $0$.

Si queremos calcular el seno de 360° directamente fallaremos:

In [23]:
math.sin(360)

0.9589157234143065

El problema es que las funciones trigonométricas en Python aceptan  los grados en radianes.  Recordemos que $\pi$ radianes son 180° sexagesimales, todo lo demás es regla de tres simple:
- 360° $\to$ $2\pi$
- 270° $\to$ $\displaystyle\frac32 \pi$
- 90° $\to$ $\pi/2$

En general
- $x$ grados sexagesimales $\to$ $x \cdot \pi / 180$ radianes.

Podemos obtener $\pi$ copiando alguna estimación de algún sitio web, pero es más sencillo utilizar el que nos provee  Python:

In [24]:
math.pi

3.141592653589793

In [25]:
math.sin(2 * math.pi)

-2.4492935982947064e-16

La expresión resultante es

$$-2.4492935982947064e-16$$
 que se lee

$$
-2.4492935982947064 \times 10^{-16}.
$$
 Es decir es

$$
-0.00000000000000024492935982947064
$$
Nosotros esperaríamos que el resultado de ejecutar la línea de código anterior sea `0`, pero debido a que el valor de $\pi$ es una aproximación y  que la función `sin()` también devuelve aproximaciones,  el resultado es número "cercano"  a `0`.

Para ver todas las funciones del módulo ejecutamos `dir(math)`:

In [26]:
dir(math)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'lcm',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'nextafter',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc',
 'ulp']

Entre todos los métodos que nos provee la biblioteca `math` tenemos los que convierten radianes a grados sexagesimales y  viceversa.

In [27]:
math.degrees(math.pi)

180.0

In [28]:
math.sin(math.radians(360)) # esto debería resultar un número cercano a 0

-2.4492935982947064e-16

También podemos calcular exponenciales y logaritmos.

In [29]:
print(math.e) # el número e
math.log(math.e) # el logaritmo natural (en base e) de e

2.718281828459045


1.0

Recordemos que $\log(x) = y$ si y solo si $e^y = x$.

En  general
$$
\log_a(x) = y \quad\Leftrightarrow \quad a^y = x  \quad \Leftrightarrow\quad  \log(a^y) = \log(x) \quad  \Leftrightarrow \quad \log(a) \cdot y = \log(x)\quad  \Leftrightarrow \quad  y = \log(x) / \log(a)
$$

Concluyendo:
$$
\log_a(x) =  \log(x) / \log(a)
$$

Luego, si queremos calcular, el logaritmo en base 2 de `x`, debemos  hacer `math.log(x) / math.log(2)`. Por ejemplo,

In [30]:
print(math.log(10) / math.log(2)) # logartimo en base 2 de 10
print(math.log(16) / math.log(2)) # logartimo en base 2 de 16 (es 4, pues 2**4 = 16)

3.3219280948873626
4.0


Por suerte el logaritmo en base 2, el más utilizado en ciencias de la computación,  está definido en la biblioteca  `math`.

In [31]:
print(math.log2(10))
print(math.log2(16))

3.321928094887362
4.0


Una forma de elevar un número  a una potencia es utilizando `math.pow()`

In [ ]:
math.pow(2, 3)

que es lo mismo que

In [32]:
2**3

8

También, para ciertas situaciones y algoritmos a veces es conveniente manejar el concepto de infinito. Esto muchas veces se hace eligiendo un número "muy grande". Python nos provee `math.inf` que pese a no ser un número formalmente es del tipo `float`. Decimos que no es número porque no cumple las leyes o axiomas de un número real pues es un "número" más grande que cualquier otro.

Para que quede claro: matemáticamente *infinito* no es un número sino un concepto matemático.

In [33]:
type(math.inf)

float

Por  el tipo `math.inf` parece ser un número, pero...

In [34]:
math.inf + 1 == math.inf # ningún número cumple con esta condición

True

In [36]:
math.inf / 2 == math.inf # ningún número, excepto el 0, cumple con esta condición

True

El `==` compara dos cantidades y devuelve `True` si son iguales.

In [37]:
print(10**4552 < math.inf) # todos los números cumplen con esta condición

True


Para evitar escribir `math.` antes de cada función ejecutamos la siguiente celda  de código:

In [38]:
from math import *

Luego de ejecutar la celda de arriba es lícito escribir, por ejemplo:

In [39]:
sin(2 * pi)

-2.4492935982947064e-16

Si sabemos que vamos a utilizar algunas funciones determinadas del módulo, podemos importar solamente esa funciones:

In [40]:
from math import log, e
log(e)

1.0

Si queremos acortar o cambiar el nombre del módulo dentro de nuestro programa podemos hacer algo como lo siguiente:

In [41]:
import math as mt
mt.log(mt.e)

1.0

# Variables, expresiones y sentencias

## 1. Variables y asignaciones

En programación, una variable está formada por un espacio en el sistema de almacenaje (memoria principal de un ordenador) y un nombre simbólico (un identificador) que está asociado a dicho espacio. Ese espacio contiene una cantidad de información conocida o desconocida, es decir un valor.

Las variables de un programa describen el *estado* del programa. Los estados describen completamente al programa, al menos en lo que se refiere a los resultados que se obtienen y a la interacción con el mundo exterior u otras componentes.

En Python la forma de crear variables es crear un nombre, el de la variable, y asignarle un valor. Una *asignación* crea una nueva variable y por medio del símbolo `=` le da un valor:

In [42]:
mensaje = 'Y ahora algo completamente diferente'
n = 17
pi = 3.1415926535897932

Este ejemplo hace tres asignaciones. La primera asigna una cadena a una nueva variable llamada `mensaje`; la segunda asigna el número entero `17` a `n` y  la tercera asigna el valor (aproximado) de $\pi$ a `pi`.


In [43]:
print(mensaje)
print(n)
print(pi)
print(2 * pi)

Y ahora algo completamente diferente
17
3.141592653589793
6.283185307179586


## 2.  Nombres de variables y palabras reservadas
Como norma general, los programadores eligen nombres significativos para sus
variables: esto permite documentar para qué se usa la variable.

Los nombres de las variables pueden tener una longitud arbitraria. Pueden estar
formados por letras y números, pero deben comenzar con una letra o un guion bajo (`_`). Aunque es aceptable usar mayúsculas, por convención no lo haremos. Si lo hacés, recordá que la distinción es importante: `Bruno` y `bruno` son dos variables diferentes.

El guion bajo (`_`) también es legal y se utiliza a menudo para separar nombres con múltiples palabras, como `mi_nombre` o `precio_del_cafe_colombiano` o incluso el carácter simple se utiliza como variable.

Para definir una variable debemos asignarle un valor. Por ejemplo:

In [44]:
bruno_diaz = 5
bruno = 6
_ = 38
_empleado = 'Juan Pérez'


Si intentás darle a una variable un nombre ilegal, obtendrás un error de sintaxis.

In [ ]:
# 76trombones = "gran desfile" # descomentar esta linea produce error 

In [ ]:
# mas$ = 1000000 # descomentar esta linea produce error 

In [ ]:
# global = "Curso de Programación 101" # descomentar esta linea produce error

`76trombones` es ilegal porque no comienza por una letra o guion bajo. `mas$` es ilegal porque contiene un carácter ilegal, el signo pesos. Pero ¿qué tiene de malo `global`?

Resulta que `global` es una de las palabras reservadas de Python. El lenguaje usa las palabras reservadas para definir sus reglas y estructura, y no pueden usarse como nombres de variables.

Python 3 tiene las siguientes palabras reservadas (son pocas):

|      |     |     |      |     |
| :------------- | :---------- | :----------- | :----------- | :----------- |
| `False`| `class`| `finally`| `is`| `return`|
| `None`| `continue`| `for`| `lambda`| `try`|
| `True`| `def`| `from`| `nonlocal`| `while`|
| `and`| `del`| `global`| `not`| `with`|
| `as`| `elif`| `if`| `or`| `yield`|
| `assert`| `else`| `import`| `pass`|
| `break`| `except`| `in`| `raise`|

Tal vez quieras mantener esta lista a mano.

Si el intérprete se queja de alguno de sus nombres de variable, y  no sabés por qué, comprobá si está en esta lista.




## 3. Expresiones y sentencias

Una *expresión* es una línea de código válido Python que produce un valor.
Podemos pensar en las expresiones como las palabras del lenguaje.
Un valor en sí mismo se considera una expresión y, por lo tanto, una variable también es una expresión.

Por ejemplo,  si hacemos la asignación



In [45]:
n = 17


Entonces, las siguientes son todas expresiones legales:
```
42
n
n + 25
```
Cuando escribís una expresión en el prompt o en una celda de código y la ejecutás, el intérprete la evalúa, lo que significa que encuentra el valor de la expresión. En este ejemplo, en la primer línea el intérprete sabe que hay un valor `42`.  En la tercera línea  `n` tiene el valor `17` y en la última `n + 25` tiene el valor `42`.

Una *sentencia*, también llamada *declaración* o en inglés *statement*, realiza una acción. Escribimos una sentencia cuando queremos que nuestro programa haga algo. Es decir, las sentencias son las instrucciones de nuestro programa.

En este Colab hemos visto dos tipos de sentencias: `print` y la asignación.

In [46]:
print(1)
x = 2
print(x)

1
2


Cuando escibís una sentencia en la linea de comandos, Python la ejecuta
y muestra el resultado (si lo hay). En  el caso de Jupyter Notebooks, solo se muestra el valor de la última sentencia en la secuencia.

Las sentencias de asignación no entregan ningún resultado.



In [47]:
millas = 26.2
millas * 2

52.4

## 4. Operaciones de cadena

Recordemos que las _cadenas_, o _strings_ en inglés, representan porciones de texto. Toda cadena es una serie de símbolos alfanuméricos sucesivos delimitados por  comillas simples o dobles.

Hablaremos en esta sección de que tipo de operaciones se pueden hacer con las cadenas.

En general, no es posible realizar operaciones matemáticas con cadenas, incluso si las cadenas parecen números, por lo que lo siguiente es ilegal:


```
'2' * '3'  
'huevos' / 'fácil'  
'tercero' * 'un amuleto'
```
Sí podemos utilizar los operadores `+` y `*` con cadenas, pero no son operadores aritméticos. Lo vemos a continuación.

El operador `+` realiza la concatenación de cadenas, lo que significa que une las cadenas vinculándolas de un extremo a otro. Por ejemplo:


In [48]:
primero = 'sobre'
segundo = 'todo'
print(primero + segundo)

sobretodo



Como dijimos arriba, el operador `*` también trabaja con cadenas; realiza repetición. Por ejemplo, `'Spam' * 3` es `'SpamSpamSpam'`. Si uno de los valores es una cadena, el otro debe ser un número entero.



In [49]:
print('Spam' * 3)
print(2 * 'Spam')

SpamSpamSpam
SpamSpam


Este uso de `+` y `*` tiene sentido por analogía con la suma y la multiplicación. Así como `4 * 3` es equivalente a `4 + 4 + 4`, esperamos que `'Spam' * 3` sea lo mismo que `'Spam' + 'Spam' + 'Spam'`, y lo es. Hay muchas propiedades de la suma y multiplicación de enteros  que se aplican a la suma y multiplicación de cadenas. Por ejemplo, las asociatividad de la suma vale para ambos contextos.  

También existe un elemento neutro de la suma y  es la _cadena vacía_ `''`.

Por otro lado, la conmutatividad no es una propiedad que vale para la concatenación de cadenas ¿Podés pensar en otra propiedad que tenga la suma de enteros y que no tenga la concatenación de cadenas?


## 5. Comentarios de Python

A medida que los programas se hacen más grandes y complicados, se vuelven más difíciles de leer. Los lenguajes formales son densos y, a menudo, es difícil mirar un fragmento de código y averiguar qué está haciendo o por qué.

Por esta razón, es una buena idea agregar anotaciones a tus programas para explicar en lenguaje natural lo que está haciendo el programa. Estas anotaciones o notas se denominan _comentarios_ y en Python comienzan con el símbolo `#`:
```
# calcula el porcentaje de la hora que ha transcurrido
porcentaje = (minuto * 100) / 60
```
En este caso, el comentario aparece en una línea aparte. También se pueden poner comentarios al final de una línea:

```
porcentaje = (minuto * 100) / 60 # porcentaje de una hora
```

Todo, desde el `#` hasta el final de la línea, se ignora; no tiene ningún efecto en la ejecución del programa.

Los comentarios son más útiles cuando documentan características no obvias del código. Es razonable suponer que el lector del código puede averiguar _qué_ hace  y es útil explicar _por qué._

Este comentario es redundante e inútil:
```
v = 5 # asignar 5 a v
```

Este comentario contiene información útil que no está en el código:
```
v = 5  # velocidad en metros / segundo.
```

Los buenos nombres de variables pueden reducir la necesidad de comentarios, pero los nombres largos pueden dificultar la lectura de expresiones complejas, por lo que se debe ser cuidadoso y usar buen criterio para darle nombres a las variables.



## 6. Depuración

Pueden ocurrir tres tipos de errores en un programa: errores de sintaxis, errores de tiempo de ejecución y errores semánticos. Es útil distinguirlos para localizarlos más rápidamente.

**Error de sintaxis:** *Sintaxis* se refiere a la estructura de un programa y las reglas sobre esa estructura. Por ejemplo, los paréntesis deben venir en pares coincidentes, por lo que `(1 + 2)` es legal, pero `8)` es un error de sintaxis.

Si hay un error de sintaxis en cualquier parte de tu programa, Python muestra un mensaje de error y se cierra, y no podrá ejecutar el programa.

Durante las primeras semanas de aprendizaje en programación, es posible que uno dedique mucho tiempo a rastrear errores de sintaxis,  aunque los entornos de programación a menudo nos indican inmediatamente el error y lo corregimos.

A medida que se gane experiencia, se cometerán menos errores y se los encontrará más rápido.

**Error de tiempo de ejecución:** el segundo tipo de error es un *error de tiempo de ejecución*, llamado así porque el error no aparece hasta que el programa ha comenzado a ejecutarse. Estos errores también se denominan *excepciones* porque suelen indicar que ha sucedido algo excepcional (y malo).

Los errores de tiempo de ejecución son raros en los programas simples que veremos en las primeras clases, por lo que puede pasar un tiempo antes de que encontrés uno.

**Error semántico:** el tercer tipo de error es *semántico* que significa relacionado con el significado.

Si hay un error semántico en tu programa, se ejecutará sin generar mensajes de error, pero no hará lo correcto. Hará otra cosa. Específicamente, hará lo que le dijiste que hiciera (que puede no coincidir con lo que querés que haga).

Identificar errores semánticos puede ser complicado porque requiere trabajar hacia atrás mirando la salida del programa y tratando de averiguar qué está haciendo.




In [50]:
# imprimamos el valor 4,5
print(4,5) # error semántico

4 5


In [51]:
print(4.5) # correcto

4.5


Tanto para los errores sintácticos como para las excepciones los entornos de programación nos ofrecen mucha ayuda, incluso marcándolos antes de ejecutar el programa. Sin embargo,  los errores semánticos solo los podemos solucionar  razonando sobre lo que hemos escrito en el programa.  